File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/typing/composing/TypingUndoManager.kt
package com.lobster.input.keyboard.typing.composing

/**
 * 打字模式撤销栈:记录最近上屏文本,支持一键撤销(与语音模式 undo 独立)。
 */
class TypingUndoManager(private val maxDepth: Int = 32) {

    private val stack = ArrayDeque<String>()

    fun record(text: String) {
        if (text.isEmpty()) return
        stack.addLast(text)
        while (stack.size > maxDepth) stack.removeFirst()
    }

    fun canUndo(): Boolean = stack.isNotEmpty()

    fun peek(): String? = stack.lastOrNull()

    fun pop(): String? = if (stack.isEmpty()) null else stack.removeLast()
}

File: lobster-input-ios/LobsterKeyboard/Typing/Composing/TypingUndoManager.swift
import Foundation

/// 打字模式撤销栈:记录最近上屏文本,支持一键撤销(与语音模式 undo 独立)。
final class TypingUndoManager {

    private let maxDepth: Int
    private var stack: [String] = []

    init(maxDepth: Int = 32) {
        self.maxDepth = maxDepth
    }

    func record(_ text: String) {
        guard !text.isEmpty else { return }
        stack.append(text)
        while stack.count > maxDepth { stack.removeFirst() }
    }

    func canUndo() -> Bool { !stack.isEmpty }

    func peek() -> String? { stack.last }

    func pop() -> String? {
        guard !stack.isEmpty else { return nil }
        return stack.removeLast()
    }
}
